diff --git a/tensorflow/lite/CMakeLists.txt b/tensorflow/lite/CMakeLists.txt
index 0924ca9d..812f8c55 100644
--- a/tensorflow/lite/CMakeLists.txt
+++ b/tensorflow/lite/CMakeLists.txt
@@ -564,14 +564,17 @@ endif()
 set(_ALL_TFLITE_SRCS
   ${TFLITE_CORE_ACCELERATION_SRCS}
   ${TFLITE_CORE_API_SRCS}
-  ${TFLITE_CORE_C_SRCS}
+
+  ${TFLITE_SOURCE_DIR}/c/common_internal.cc
+  ${TFLITE_SOURCE_DIR}/c/c_api_opaque_internal.cc
+  ${TFLITE_SOURCE_DIR}/core/c/c_api_opaque.cc
+
   ${TFLITE_CORE_EXPERIMENTAL_SRCS}
   ${TFLITE_CORE_KERNELS_SRCS}
   ${TFLITE_CORE_SRCS}
   ${TFLITE_CORE_ASYNC_SRCS}
   ${TFLITE_CORE_ASYNC_C_SRCS}
   ${TFLITE_CORE_TOOLS_SRCS}
-  ${TFLITE_C_SRCS}
   ${TFLITE_DELEGATES_FLEX_SRCS}
   ${TFLITE_DELEGATES_GPU_SRCS}
   ${TFLITE_DELEGATES_NNAPI_SRCS}
diff --git a/tensorflow/lite/c/CMakeLists.txt b/tensorflow/lite/c/CMakeLists.txt
index 3fb086ec..f91a0fe4 100644
--- a/tensorflow/lite/c/CMakeLists.txt
+++ b/tensorflow/lite/c/CMakeLists.txt
@@ -86,5 +86,7 @@ if (TFLITE_C_BUILD_SHARED_LIBS)
 endif()
 
 target_link_libraries(tensorflowlite_c
+  -Wl,--whole-archive
   tensorflow-lite
+  -Wl,--no-whole-archive
 )
